iree/compiler/Conversion/LinalgToSPIRV/SplitDispatchFunctionPass.cpp

```diff
@@ -57,35 +58,67 @@ namespace iree_compiler {
 
 namespace {
 
-/// Returns true if the Linalg ops can be separated to multiple kernels.
-bool canSeparateOps(ArrayRef<Operation *> ops) {
-  if (llvm::any_of(ops, [](Operation *op) {
-        if (auto linalgOp = dyn_cast<linalg::LinalgOp>(op))
-          return !linalgOp.hasBufferSemantics();
-        return false;
-      }))
-    return false;
-
-  // Require no other non-metadata ops interleave with Linalg structured ops for
-  // now. This is the common case and it simplifies further analysis.
+/// For the list of operations in `ops` returns a list of lists where each list
+/// contains the operations that need to be put in a separate dispatch function.
+LogicalResult separateOps(
+    ArrayRef<Operation *> ops,
+    const linalg::LinalgDependenceGraph &dependenceGraph,
+    SmallVectorImpl<SmallVector<Operation *, 1>> &fusedOpList) {
+  assert(!ops.empty() &&
+         "expected at least one separable op for splitting dispatch function");
+  SmallVector<Operation *, 1> currList;
   for (auto currOp = ops.begin(), nextOp = std::next(ops.begin());
        nextOp != ops.end(); ++currOp, ++nextOp) {
+    // Check that the operation has buffer semantics.
+    if (auto linalgOp = dyn_cast<linalg::LinalgOp>(*currOp)) {
+      if (!linalgOp.hasBufferSemantics()) return failure();
+    }
+
+    // Require no other non-metadata ops interleave with Linalg structured ops
+    // for now. This is the common case and it simplifies further analysis.
     Operation *iter = (*currOp)->getNextNode();
     while (iter != *nextOp && (MemoryEffectOpInterface::hasNoEffect(iter) ||
                                isa<IREE::PlaceholderOp>(iter)))
       iter = iter->getNextNode();
-    if (iter != *nextOp) return false;
-  }
+    if (iter != *nextOp) return failure();
+
+    currList.push_back(*currOp);
+
+    // If the nextOp is not fusible with the currOp, then record the list of ops
+    // so far, and start a new list.
+    if (isa<linalg::LinalgOp>(*currOp) && isa<linalg::LinalgOp>(*nextOp) &&
+        dependenceGraph.hasDependenceFrom(
+            *currOp, *nextOp,
+            linalg::LinalgDependenceGraph::DependenceType::WAW)) {
+#define ADD_FUSABLE_PAIR(SrcOpTy, DstOpTy) \
+  if (isa<SrcOpTy>(*currOp) && isa<DstOpTy>(*nextOp)) continue;
+
+      ADD_FUSABLE_PAIR(linalg::FillOp, linalg::ConvOp)
```

Comment:
Can you document where this list should be kept in sync with (`LinalgTileAndFusePass.cpp`) at both places? Otherwise it's quite mysterious and we may forget to update even after enabling more fusion cases.

Reply:
Done. Moved the logic of deciding when to fuse into a separate function.